src/nm-core-utils.c

```diff
@@ -2308,60 +2308,142 @@ nm_utils_is_specific_hostname (const char *name)
 
 /*****************************************************************************/
 
-gboolean
-nm_utils_machine_id_parse (const char *id_str, /*uuid_t*/ guchar *out_uuid)
-{
-	int i;
-	guint8 v0, v1;
-
-	if (!id_str)
-		return FALSE;
-
-	for (i = 0; i < 32; i++) {
-		if (!g_ascii_isxdigit (id_str[i]))
-			return FALSE;
-	}
-	if (id_str[i] != '\0')
-		return FALSE;
+typedef struct {
+	NMUuid bin;
+	char _nul_sentinel; /* just for safety, if somebody accidentally uses the binary in a string context. */
 
-	if (out_uuid) {
-		for (i = 0; i < 16; i++) {
-			v0 = g_ascii_xdigit_value (*(id_str++));
-			v1 = g_ascii_xdigit_value (*(id_str++));
-			out_uuid[i] = (v0 << 4) + v1;
-		}
+	/* depending on whether the string is packed or not (with/without hyphens),
+	 * it's 32 or 36 characters long (plus the trailing NUL).
+	 *
+	 * The difference is that boot-id is a valid RFC 4211 UUID and represented
+	 * as a 36 ascii string (with hyphens). The machine-id technically is not
+	 * a UUID, but just a 32 byte sequence of hexchars. */
+	char str[37];
+	bool is_fake;
+} UuidData;
+
+static UuidData *
+_uuid_data_init (UuidData *uuid_data,
+                 gboolean packed,
+                 gboolean is_fake,
+                 const NMUuid *uuid)
+{
+	nm_assert (uuid_data);
+	nm_assert (uuid);
+
+	uuid_data->bin = *uuid;
+	uuid_data->_nul_sentinel = '\0';
+	uuid_data->is_fake = is_fake;
+	if (packed) {
+		G_STATIC_ASSERT_EXPR (sizeof (uuid_data->str) >= (sizeof (*uuid) * 2 + 1));
+		_nm_utils_bin2hexstr_full (uuid,
+		                           sizeof (*uuid),
+		                           '\0',
+		                           FALSE,
+		                           uuid_data->str);
+	} else {
+		G_STATIC_ASSERT_EXPR (sizeof (uuid_data->str) >= 37);
+		_nm_utils_uuid_unparse (uuid, uuid_data->str);
 	}
-	return TRUE;
+	return uuid_data;
 }
 
-char *
-nm_utils_machine_id_read (void)
+/*****************************************************************************/
+
+static const UuidData *
+_machine_id_get (void)
 {
-	gs_free char *contents = NULL;
-	int i;
+	static const UuidData *volatile p_uuid_data;
```

Comment:
why volatile, is it to make the function thread-safe? Since NM doesn't use threads I think we can ignore this problem

Reply:
The code is not only used by NetworkManager binary, but also tests, nm-initrd-generator, nm-iface-helper, etc. These are all single-threaded (are they?). Note that glib does start multiple threads and GDBus is multithreadded API, however we are not supposed to use this function outside of the main thread. True.

These utility functions are at the bottom of the call-stack. I don't mind that NMManager is not thread-safe, because clearly it would be complicated, inefficient and unnecessary even to do so. But having lower layer function not thread-safe, seems very undesirable to me (out of principle). Code in `shared/nm-utils` and `src/nm-core-utils.c` and `libnm-core/nm-utils.c` should just be thread-safe.

Most of such singleton-patterns only require that we initialize *one* instance (and in case of races to to throw away the looser of the race). I wish, for such a trivial case we would find a preferred wat to implement it, which is
 1) thread-safe
 2) with little performance impact
 3) with little readability impact (simplicity)


About 1), I am indeed not entirely sure whether this is thread safe on all architectures that we support. I *think* it is, is it not? Same for `generate_duid_from_machine_id()`. I think that is actualyl thread-safe, isn't it?

Reply:
> The code is not only used by NetworkManager binary, but also tests, nm-initrd-generator, nm-iface-helper, etc. These are all single-threaded (are they?). Note that glib does start multiple threads and GDBus is multithreadded API, however we are not supposed to use this function outside of the main thread. True.

I agree.

> But having lower layer function not thread-safe, seems very undesirable to me (out of principle). Code in `shared/nm-utils` and `src/nm-core-utils.c` and `libnm-core/nm-utils.c` should just be thread-safe.

In my opinion, since those functions are part of the NM library of utilities and none of the NM components is multithreaded, making those functions thread-safe just adds a bit of unneeded complexity. But anyway, I don't mind too much.

Regarding the current implementation, I think the 'volatile' keyword is not needed since we are always using dedicated functions (with internal memory barriers) to access the variable.

Reply:
> Regarding the current implementation, I think the 'volatile' keyword is not needed since we are 
> always using dedicated functions (with internal memory barriers) to access the variable.

You are right. To my understanding, volatile only matters when you read/write the variable. But these variables are never directly accessed. All reads/writes happens like `g_atomic_pointer_get(&variable)`, which already casts the variable's address as volatile. I think declaring the variable as volatile has not effect in practice, but it seems it bit more correct since we indeed always access the variable by taking a (volatile) pointer to it. That's why I would keep it.